File: backend/services/renderer.py
from __future__ import annotations

from functools import lru_cache
from pathlib import Path
from typing import Any

from jinja2 import Environment, FileSystemLoader, select_autoescape

_TEMPLATE_DIR = Path(__file__).resolve().parent.parent / "templates"


@lru_cache(maxsize=1)
def _env() -> Environment:
    return Environment(
        loader=FileSystemLoader(str(_TEMPLATE_DIR)),
        autoescape=select_autoescape(["html"]),
        trim_blocks=True,
        lstrip_blocks=True,
    )


def render_lesson_plan(
    *,
    lesson: dict[str, Any],
    cluster: dict[str, Any],
    plan_json: dict[str, Any],
    knowledge_bases_used: list[dict[str, Any]],
    provider: str,
    model_used: str,
) -> str:
    tpl = _env().get_template("lesson_plan.html.j2")
    return tpl.render(
        lesson=lesson,
        cluster=cluster,
        recommendations=plan_json.get("recommendations") or [],
        plan_steps=plan_json.get("plan_steps") or [],
        companion_materials=plan_json.get("companion_materials") or [],
        knowledge_bases_used=knowledge_bases_used,
        provider=provider,
        model_used=model_used,
    )

File: server/src/services/renderer.js
const ejs = require("ejs");
const path = require("path");

const TEMPLATE_DIR = path.resolve(__dirname, "..", "templates");

async function renderLessonPlan({ lesson, cluster, planJson, knowledgeBasesUsed, provider, modelUsed }) {
  return ejs.renderFile(
    path.join(TEMPLATE_DIR, "lesson_plan.ejs"),
    {
      lesson,
      cluster,
      recommendations: planJson.recommendations || [],
      plan_steps: planJson.plan_steps || [],
      companion_materials: planJson.companion_materials || [],
      knowledge_bases_used: knowledgeBasesUsed,
      provider,
      model_used: modelUsed,
    },
    { async: false }
  );
}

module.exports = { renderLessonPlan };
